scanner/decisions/0012-reindexing.md

```diff
@@ -0,0 +1,192 @@
+and re-index the image. As mentioned in the [Context](#context) section, ClairCore already has some builtin performance optimizations for this.
+Another benefit to this that `*claircore.Manifest` continues to only be generated, as needed.
+
+However, there are cons which outweigh the pros:
+
+* More data for Scanner V4 to track outside ClairCore
+  * This is not a blocker, but it is worth mentioning
+* Handling preexisting Index Reports
+  * This is the challenging aspect, which would block this
+  * There are a few options:
+    * Write a migration
+      * This involves doing essentially the same thing `ManifestScanned` does to determine the "versioned scanners"
+      * This may not be able to be pure PostgeSQL, as the `etag` is generated in Go, so that is an extra challenge
+    * Populate the table upon Index Report fetching
+      * If the entry does not exist in the table (the case for all Index Reports created prior to this release), then create the Index Report and populate the table
+      * This essentially forces a re-indexing of all previously indexed images, whether it is required or not, which may be costly
+
+### Random Deletions
+
+When it comes to variables outside "versioned scanners" there is no currently known clean way to handle the re-indexing.
+
+Something like creating our own "versioned scanner" to track other types of changes is not a solution:
+
+* Updating the version of the custom "versioned scanner" will only affect that scanner, and not the others which would all be skipped upon re-indexing
+* It is not feasible to track every single change in ClairCore which may affect Index Reports
+
+Instead, we will randomly delete manifests from the ClairCore database. Deleting a manifest from ClairCore
+deletes all data related to it, including the Index Report.
+Doing this solves both the remaining aspects of the re-indexing problem and the database ever-growing problem,
+as unneeded Index Reports are permanently deleted and needed reports may just be regenerated.
+
+**Note**: this approach also solves the "versioned scanner" update scenario; however, we opt to implement the [`checkManifest`](#reuse-claircores-checkmanifest-step)
+step for faster turnaround upon "versioned scanner" updates.
+
+Of course, determining a timeframe for deletion is non-trivial. Deleting too quickly means Index Reports may be corrected
+faster, but too many (probably) unnecessary requests to the image registry and (probably) unnecessary CPU/memory/disk usage.
+Deleting too slowly means resource usage will be much less, but it may take (too) long to correct the Index Report.
+Similarly, we would need to ensure not all Index Reports are all re-indexed at the same time (hence random deletions)
+to ensure spikes are minimized.
+
+The deletions will be chosen from a range between **one week** and **one month**. We believe this is a sufficient timeframe
```

Comment:
How will the time be chosen? Configuration/env var?

Reply:
random. The current PR does this:

```
// randomExpiry generates a random time.Time between seven and thirty days from now
func randomExpiry(now time.Time) time.Time {
	// now + a random number of seconds between zero and twenty-three days + seven days
	expirySec := now.Unix() + rand.Int64N(twentyThreeDays) + sevenDays
	return time.Unix(expirySec, 0)
}
```

and also:

```
// insertMissingManifests inserts missing manifests from the manifest table into manifest_metadata,
// and it sets the expiration time to some random time in the future between seven days from now
// and thirty days from now.
const insertMissingManifests = `
	INSERT INTO manifest_metadata (manifest_id, expiration)
	SELECT m.hash, now() + (make_interval(days => 23) * random()) + make_interval(days => 7)
	FROM manifest m
	WHERE NOT EXISTS (
		SELECT FROM manifest_metadata mm WHERE mm.manifest_id = m.hash
	)
	RETURNING manifest_id`
```